Satisfaction implementation

Rinnerdos opened the pull request:
Deleted the previous branch, and made a new one. For the structural customer constructor, I'm not sure how to add satisfaction in the object.

Rinnerdos committed:
See PR

N-Wouda committed:
Fix customer update satisfaction

Rinnerdos committed:
Satisfaction begin

Deleted the previous branch, and made a new one. For the structural customer constructor, I'm not sure how to add satisfaction in the object.

Rinnerdos committed:
Customer satisfaction

Added updateSatisfaction, topbar satisfaction and emojis.

Rinnerdos committed:
Player and customer satisfaction implementation

Adds the player and cusotmer satisfaction, which are tied to score. Score and satisfaction is controlled in the satisfaction controller. Tomorrow I'll finish the smiley bit.

Rinnerdos commented on src/controllers/moneycontroller.js:
```diff
@@ -11,7 +11,10 @@ export default class MoneyController extends Controller
     {
-        GAME.model.config.money = GAME.model.config.money + parseFloat(amount);
-        $("#money").html(GAME.model.config.money.toFixed(2));
+        const config = GAME.model.config;
```
I changed this because first I also updated score here (which still might be handy). Therefore it was useful to use the constant config, now it isn't necessary per se but forgot to change it back.

Rinnerdos commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,61 @@
+        });
+    }
+
+    /**
```
As this says; score is updated every hour. This might not be ideal since it also should be updated when changes in money and/or player satisfaction happen. We could discuss this tomorrow.

Rinnerdos committed:
Emojis

Adds the 32 and 64 bit version of the emojis to the assets, and tries to implement the adding of said emojis to the customers. However, I'm unsure how to make each customer have its own customer-id in the html. Also for me the game still displays a ":)" after each name of the customer, though this code was removed 2 commits ago.

Rinnerdos committed:
Scorecontroller

Added a scorecontroller, with the function updateScore. This function gets called after the money or playersatisfaction changes. Also put the tresholds for neutral/angry/sendAway in the config file.

Rinnerdos commented on src/controllers/customercontroller.js:
```diff
@@ -140,22 +140,18 @@ export default class CustomerController extends OrderController
         });
     }
 
-    /**
```
Since the order/customer view is also updated by the score controller, I have changed it to a 'public' function. I hope that's not a problem/should be implemented differently.

N-Wouda replied:
That should be fine, I think.

Rinnerdos committed:
Minor changes

Deleted some empty lines and removed score calculation from the satisfaction controller.

N-Wouda commented on src/controllers/moneycontroller.js:
```diff
@@ -11,7 +12,13 @@ export default class MoneyController extends Controller
-        $("#money").html(GAME.model.config.money.toFixed(2));
+        let config = GAME.model.config;
+
+        const scoreController = new ScoreController();
```
Do this after updating the money (this should ideally be the line directly before `updateScore`).

N-Wouda commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,50 @@
+import ScoreController from "./scorecontroller";
+
+export default class SatisfactionController extends Controller
+
```
This white space is not needed.

N-Wouda commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,50 @@
+    {
+        let config = GAME.model.config;
+
+        const scoreController = new ScoreController();
```
This thing may be moved down in the method to nearly the end.

N-Wouda replied:
Initialize variables where you use them, _not_ before.

N-Wouda commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,50 @@
+
+            let config = GAME.model.config;
+
+            if (customer.satisfaction <= config.sendAwayThreshold)
```
Compound statement needs brackets.

N-Wouda commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,50 @@
+        GAME.model.customers.forEach(function (customer) {
+            customer.satisfaction -= 10;
+
+            const customerController = new CustomerController();
```
Only used in the if statement, so init it there (block scoping is better than function-level globals).

Rinnerdos replied:
Yeah, I still have the tendency to init things all at the same place/beginning.

Rinnerdos committed:
Minor changes

I believe I have improved everything you commented on Niels. The compound statement not having brackets already was giving some bugs, so thanks for reminding me haha. Now the cusotmers that are sent away will also count in the player satisfaction.

Rinnerdos commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -31,16 +30,19 @@ export default class SatisfactionController extends Controller
-            if (customer.satisfaction <= config.sendAwayThreshold)
+            if (customer.satisfaction <= config.sendAwayThreshold) {
+                const customerController = new CustomerController();
+                const satisfactionController = new SatisfactionController();
```
I made a new SatisfactionController, even though this is inside the satisfaction controller. So Maybe this should be done otherwise? Didn't know how though.

N-Wouda replied:
Use `this`.

`this.updatePlayerSatisfaction(customer)`.

Rinnerdos replied:
Yeah I tried that but then it gets buggy and gives the error:

Uncaught TypeError: Cannot read property 'updatePlayerSatisfaction' of undefined

N-Wouda replied:
It's not buggy. This happens because you still use `function` syntax, which declares a function-level `this` argument.

Moving to ES6 arrow syntax should resolve this.

Rinnerdos replied:
Ahhh alright :D

N-Wouda commented:
@Rinnerdos you may update this branch from `/dev`, and then resolve the merge conflicts. This should give you the latest panel for the customer view, into which the images for the satisfaction can be inserted.

Rinnerdos committed:
Function changed to implement ES6

Rinnerdos committed:
Merge branch 'dev' into satisfaction

Rinnerdos commented:
I'm struggling with the .data/selecting the right image. I can access the customer.id if I store it as customer-data or whatever, though this doesn't solve which element I have to target with jquery right? I tried it with the following code (probably bad, but alas) 

`  if ($("#emoji").data("customer") == customer.id)
                {
                    $("#emoji").attr("src", "src/assets/img/emojis/neutralSmall.png");
                }
`
I'll push what I have but it doesn't work (I mean, #emoji will only select the first image it finds to that makes it extra wrong I guess). Why can't I just make an id="{{ id }}" , and target that?

Rinnerdos committed:
Small updates and img insert

Not working, see comment earlier.

Rinnerdos commented on src/views/template/topbar.html:
```diff
@@ -31,7 +30,7 @@
             <span id ="score">{{ config.score }}</span>
```
Still, only the score displays. Not "Score"  itself.

N-Wouda replied:
Do you have the text anywhere? I think this is just the number.

Also, `id="score"` (alignment)

N-Wouda commented:
Sure you do, once you know the element, the image should be one of its children (select the element by customer-data id, and then use its child `img`).

Or something along those lines. @lukko can you help @Rinnerdos resolve this?

N-Wouda commented on src/views/template/customer/customerorder.html:
```diff
@@ -3,6 +3,9 @@
+        <span class="image">
+            <img id="{{ id }}" src="src/assets/img/emojis/happySmall.png" alt="Emoji" data-emoji="{{ id }}">
+        </span>
         <button type="button" class="customer-serve btn btn-default btn-xs" style="float: right;" data-customer="{{ id }}">
```
Move `data-customer` up one level, and then target the child image.

N-Wouda commented on src/views/template/topbar.html:
```diff
@@ -21,14 +21,17 @@
-            <strong class="text-capitalize">{{#l}}satisfaction{{/l}}</strong>
-            <span id="satisfaction">{{ config.satisfaction }}</span>%
+        <p class="navbar-text">
+            <strong>{{#1}}Score{{/1}}</strong>
```
Is this a `1`, or an `l`? It should be an `l`.

Rinnerdos replied:
This explains a lot. That's the translate wrap right? Satisfaction also should be translated I guess?

N-Wouda replied:
Yes. `l` for **l**anguage.

N-Wouda commented:
@Rinnerdos how's this coming along?

Rinnerdos committed:
Small changes

Rinnerdos commented:
Still don't know how to use the data/target it. @lukko any suggestions?

N-Wouda committed:
Merge remote-tracking branch 'LCfP/satisfaction' into satisfaction

N-Wouda committed:
Update emoji views

N-Wouda committed:
TODO voor Rinnert

Rinnerdos committed:
Changes to html/css

Tried to make the customer order view consist of two columns inside one row. Left one for the emoji, right one for the rest. Unfortunately my HTML/CSS doesn't seem to update so I can't check wether it works.

lukko commented on src/views/css/main.css:
```diff
@@ -160,3 +160,11 @@ body {
+
+.customer-class {
+    width: 80%;
+}
```
You probably don't need those css classes. It can be done with Bootstrap out of the box.
```html
<li class="list-group-item">
  <div class="row">
    <div class="col-xs-3">
      <!-- emoji -->
    </div>
    <div class="col-xs-9">
      <!-- other stuff -->
    </div>
  </div>
</li>
```

Rinnerdos replied:
Ahh yeah, been a while since I did this. Thanks

lukko commented on src/controllers/satisfactioncontroller.js:
```diff
@@ -0,0 +1,47 @@
+                this.updatePlayerSatisfaction(customer);
+                customerController.sendAway(customer);
+            }
+        }
```
Where do you now determine which emoji has to be loaded? It seems as the default emoji is not changed anywhere (which might make it look like your view is not updating). Or am I overlooking something?

Rinnerdos replied:
No this should be done by the update customer view in the customer controller, then base the template to be loaded on the customer satisfaction. Not sure how to work with that though. If you have pointers/links to how to work with that I can hopefully finish it tomorrow. Shouldn't be too difficult, so maybe you could even do it. Of course, you may not have the time either.

lukko commented on src/controllers/simulationcontroller.js:
```diff
@@ -3,7 +3,7 @@ import DemandController from './demandcontroller';
 import WarehouseController from './warehousecontroller';
 import LevelController from './levelcontroller';
-
+import SatisfactionController from "./satisfactioncontroller";
```
Single quotes here.

lukko committed:
make emojis work

lukko committed:
small fixes

lukko committed:
Add line breaks

lukko commented:
Customers now look like this
<img width="333" alt="schermafbeelding 2018-03-18 om 15 06 02" src="https://user-images.githubusercontent.com/7871887/37566798-4037d26a-2abe-11e8-8f09-cb608bd66981.png">

lukko commented:
I think this is done now. Could you review it?

N-Wouda commented:
Not completely what I had in mind, but yes, this is close enough.

I'll go through the code now. @Rinnerdos @lukko.

N-Wouda commented on src/controllers/scorecontroller.js:
```diff
@@ -0,0 +1,13 @@
+
+export default class ScoreController extends Controller 
+{
+    updateScore()
```
I think this can be static.

N-Wouda replied:
(that'd solve the whole `(new ScoreController).updateScore()` stuff).

Just leave it for now, though.

Rinnerdos commented:
Seems good to me, maybe the template is not entirely what @N-Wouda meant though. Thanks for the help @lukko :)

N-Wouda commented:
I can live with all of this. Let's keep it this way, and consider this iteration finished.

Next up: move it to the server 👍 .

lukko commented:
@Rinnerdos That's right, the template is indeed a bit different than @N-Wouda's proposal. But that version was not very pleasant to look at. The emoji took too much attention, which didn't feel right.

N-Wouda commented:
@lukko I can imagine that could happen :). In any case, this is fine as-is.

Next time, you might want to post a screenshot of the implementation, and then change it (so we can come to the same conclusion). That shows you've tried the initial idea, the idea proved crap, and you came up with something better!
